### PR TITLE
Wall following demo

### PR DESCRIPTION
**Pull request by knmcguire**
Added a wall following demo as app layer example.

### CODE REVIEW

File: examples/demos/app_wall_following_demo/src/wall_following.c [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+      float height_cmd = height_sp - up_o/1000.0f;
+
+      vel_x_cmd = 0.0f;
+      vel_y_cmd = 0.0f;
+      vel_w_cmd = 0.0f;
+      float vel_w_cmd_convert=0.0f;
+
+      // Only go to the state machine if the crazyflie has reached a certain height
+      if (heightEstimate > height_sp - 0.1f)
+      {
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wall_following.c:177]: nit: inconsistent brace style here.

```
if ()
{
```

vs

```
if () {
```
elsewher.

File: examples/demos/app_wall_following_demo/src/wall_following.c [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      if (height_cmd < height_sp - 0.2f) {
+        state = stopping;
+        DEBUG_PRINT("X\n");
+      }
+
+    } else {
+      
+      // Handeling locking and unlocking
+
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wall_following.c:202]: nit: excessive spacing

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+{
+
+  if (number > (float)M_PI) {
+    return (number - (float)(2 * M_PI));
+  } else if (number < (float)(-1 * M_PI)) {
+    return (number + (float)(2 * M_PI));
+  } else {
+    return (number);
+  }
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:64]: nit: excessive newlines

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+
+// Static command functions
+static void commandTurn(float *vel_x, float *vel_w, float ref_rate)
+{
+  *vel_x = 0.0f;
+  *vel_w = direction * ref_rate;
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:73]: nit: newline

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  } else {
+    if (range > wanted_distance_from_corner) {
+      *vel_y = direction * (-1.0f * max_speed / 3.0f);
+    } else {
+      *vel_y = direction * (max_speed / 3.9f);
+    }
+    *vel_w = 0;
+  }
+
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:93]: nit: newlines

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    if (range > ref_distance_from_wall) {
+      *vel_y = direction * (-1.0f * max_speed / 3.0f);
+
+    }
+
+    else {
+      *vel_y = direction * (max_speed / 3.0f);
+
+    }
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:132]: nit: newlines

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+void adjustDistanceWall(float distance_wall_new)
+{
+  ref_distance_from_wall = distance_wall_new;
+}
+
+int wall_follower(float *vel_x, float *vel_y, float *vel_w, float front_range, float side_range, float current_heading,
+                  int direction_turn)
+{
+
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:159]: nit: newlines

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    // check if wall is found
+    bool side_range_check = side_range < ref_distance_from_wall / (float)cos(0.78f) + 0.2f;
+    bool front_range_check = front_range < ref_distance_from_wall / (float)cos(0.78f) + 0.2f;
+    if (side_range_check && front_range_check) {
+      previous_heading = current_heading;
+      angle = direction * (1.57f - (float)atan(front_range / side_range) + 0.1f);
+      state = transition(4); // go to turn_to_allign_to_wall
+    }
+    if (side_range < 1.0f && front_range > 2.0f) {
+      //  around_corner_first_turn = true;
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:206]: remove?

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+      around_corner_go_back = false;
+      previous_heading = current_heading;
+      state = transition(8); // go to rotate_around_wall
+    }
+  } else if (state == 4) { //TURN_TO_ALLIGN_TO_WALL
+    bool allign_wall_check = logicIsCloseTo(wraptopi(current_heading - previous_heading), angle, 0.1f);
+    if (allign_wall_check) {
+      // prev_side_range = side_range;
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:214]: remove?

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+      state = transition(5);
+    }
+  } else if (state == 5) {  //FORWARD_ALONG_WALL
+
+    // If side range is out of reach,
+    //    end of the wall is reached
+    if (side_range > ref_distance_from_wall + 0.3f) {
+      //  around_corner_first_turn = true;
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:222]: remove?

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+  } else if (state == 8) {   //FIND_CORNER
+    if (side_range <= ref_distance_from_wall) {
+      state = transition(6);
+    }
+
+  }
+
+  else {
+  }
+
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:254]: nit: newlines

File: examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    } else { // then turn again
+      commandTurn(&temp_vel_x, &temp_vel_w, max_rate);
+      temp_vel_y = 0.0f;
+    }
+
+  } else if (state == 5) {  //FORWARD_ALONG_WALL
+    commandForwardAlongWall(&temp_vel_x, &temp_vel_y, side_range);
+    temp_vel_w = 0.0f;
+
+    //commandForwardAlongWallHeadingSine(&temp_vel_x, &temp_vel_y,&temp_vel_w, side_range);
```
Comment (jonasdn) [inline examples/demos/app_wall_following_demo/src/wallfollowing_multiranger_onboard.c:293]: remove?

**Review by jonasdn — APPROVED** [review]
Looks good!

I made some nitty style comments, feel free to ignore them!

**Comment by knmcguire** [comment]
thanks @jonasdn and @krichardsson! I tried to implement most of  your comments. 

I still have to refactor some functions to be separate, but I couldn't find a C/C++ extensions that could do that automatically... unless you have any good suggestions?

**Pull request merged**